smu/geometry/bond_length_distribution.py

```diff
@@ -509,9 +683,117 @@ def add_from_sparse_dataframe(self, df_input, unbonded_right_tail_mass,
       atom_1 = ATOMIC_NUMBER_TO_ATYPE[atom_1]
       self.add(
           atom_0, atom_1, bond_type,
-          EmpiricalLengthDistribution.from_sparse_dataframe(
+          Empirical.from_sparse_dataframe(
               df, right_tail_mass, sig_digits))
 
+  def add_from_sparse_dataframe_file(self, filename,
+                                     unbonded_right_tail_mass, sig_digits):
+    """Adds distribution from a sparse dataframe in a csv file.
+
+    See sparse_dataframe_from_records for a description of the expected input
+    format.
+
+    Args:
+      filename: string, file to read
+      unbonded_right_tail_mass: right_tail_mass (as described in
+        Empirical) for the unbonded cases.
+      sig_digits: number of significant digits after the decimal point
+    """
+    with open(filename, 'r') as infile:
+      df = pd.read_csv(infile, dtype={'length_str': str})
+    self.add_from_sparse_dataframe(df, unbonded_right_tail_mass, sig_digits)
+
+  def _triples_from_atom_bond_atom_spec(self, spec):
+    if len(spec) != 3:
+      raise BondLengthParseError(spec)
+
+    atoms_a = self._ATOM_SPECIFICATION_MAP[spec[0]]
+    bonds = self._BOND_SPECIFICATION_MAP[spec[1]]
+    atoms_b = self._ATOM_SPECIFICATION_MAP[spec[2]]
+    yield from itertools.product(atoms_a, atoms_b, bonds)
+
+  def add_from_gaussians_file(self, filename, cutoff):
+    """Adds distribution by reading specs of Guassians.
+
+    The original intention is to take a specially formatted version of the table
+    from
+    Allen, F. H. et al. Tables of bond lengths determined by X-ray and neutron
+    diffraction. Part 1. Bond lengths in organic compounds.
+    J. Chem. Soc. Perkin Trans. 2 S1–S19 (1987)
+
+    The file shoudl be a csv with at least the columns:
+    * Bond: A bond specification like C=C (like add_from_string_spec) (or n/a)
+    * d: mean
+    * sigma: standard deviation
+    * n: number of observations of this type
+
+    Creates a Mixture distribution for every atom/pair/bond combo.
+
+    Args:
+      filename: file to read
+      cutoff: cutoff passed to Gaussian
+    """
+    with open(filename, encoding='utf-8-sig') as f:
+      reader = csv.DictReader(f)
+      for row in reader:
+        if row['Bond'] == 'n/a' or not row['d']:
+          continue
+        gaussian = Gaussian(float(row['d']), float(row['sigma']), cutoff)
+        for atom_a, atom_b, bond in self._triples_from_atom_bond_atom_spec(row['Bond']):
+          try:
+            atom_a_num = smu_utils_lib.ATOM_TYPE_TO_ATOMIC_NUMBER[atom_a]
+            atom_b_num = smu_utils_lib.ATOM_TYPE_TO_ATOMIC_NUMBER[atom_b]
+            mix_dist = self._atom_pair_dict[(atom_a_num, atom_b_num)][bond]
+          except KeyError:
+            mix_dist = Mixture()
+            self.add(atom_a, atom_b, bond, mix_dist)
+          mix_dist.add(gaussian, int(row['n']))
```

Comment:
self._atom_pair_dict typically contains objects derived from abstract class LengthDistribution. As things are now, self._atom_pair_dict might already contain objects derived from that type that do not necessarily have an .add() method (none is defined in abstract class LengthDistribution), so this line fails our internal type tests.

Possible fixes:
1. Declare abstract method add() in LengthDistribution and define it in all derived classes or
2. Re-initialize self._atom_pair_dict at the beginning of this method. This assumes that there is no prior data inside that directory that needs to be preserved or
3. A different fix, in case 1. or 2. do not lead to the desired behavior.

Reply:
I did a variant of (2) where I cast the contents of _atom_pair_dict to a Mixture.
If there is anything already in the dict, it will fail at runtime which is fine.